### PR TITLE
Implement support for MFA (SMS or email) when logging in. Also update dependencies (ng 5.1.0, etc).

### LINKED ISSUE #121: Requiring MFA causes client exception
Using MFA (SMS) causes the following exception to be thrown on the client when a successful username/password has been sent.

Cognito responds with the following:
```
{"ChallengeName":"SMS_MFA","ChallengeParameters":{"CODE_DELIVERY_DELIVERY_MEDIUM":"SMS","CODE_DELIVERY_DESTINATION":"+*******3480"},"Session":"***removed***"}
```
Which causes the following client side exception:
```
ERROR Error: callback.mfaRequired is not a function
    at CognitoUser.authenticateUserInternal (CognitoUser.js:318)
    at CognitoUser.js:294
    at Response.<anonymous> (CognitoUser.js:260)
    at Request.<anonymous> (request.js:364)
    at Request.callListeners (sequential_executor.js:105)
    at Request.emit (sequential_executor.js:77)
    at Request.emit (request.js:682)
    at Request.transition (request.js:22)
    at AcceptorStateMachine.runTo (state_machine.js:14)
    at state_machine.js:26
    at Request.<anonymous> (request.js:38)
    at Request.<anonymous> (request.js:684)
    at Request.callListeners (sequential_executor.js:115)
    at Request.emit (sequential_executor.js:77)
    at Request.emit (request.js:682)
```

### PR DESCRIPTION
**Pull request by amirtoole**
This resolves #121

### CODE REVIEW

**Comment by joe455** [comment]
@hellofornow I followed the same steps but getting error.

callback.mfaSetUp is not a function.

Any Idea?

**Comment by amirtoole** [comment]
Did you run `npm install` before trying this PR? I did update the amazon-cognito-identity-js dependency (taking a guess that's the problem).

**Comment by joe455** [comment]
Actually I installed latest version.My package.json below

 "amazon-cognito-auth-js": "^1.1.0",
    "amazon-cognito-identity-js": "^1.28.0",
    "bootstrap": "^3.3.7",
    "core-js": "^2.4.1",
    "font-awesome": "^4.7.0",
    "jquery": "^3.2.1",
    "metismenu": "^2.7.1",
    "ngx-facebook": "^2.4.0",
    "rxjs": "^5.4.2",
    "startbootstrap-sb-admin-2": "^3.3.7",
    "zone.js": "^0.8.14"

And error is,

core.es5.js:1020 ERROR Error: callback.mfaSetup is not a function
    at CognitoUser.authenticateUserInternal (CognitoUser.js:370)
    at CognitoUser.js:331
    at Response.<anonymous> (CognitoUser.js:293)
    at Request.<anonymous> (request.js:364)
    at Request.callListeners (sequential_executor.js:105)
    at Request.emit (sequential_executor.js:77)
    at Request.emit (request.js:683)
    at Request.transition (request.js:22)
    at AcceptorStateMachine.runTo (state_machine.js:14)
    at state_machine.js:26

**Comment by joe455** [comment]
@hellofornow I found the route cause finally.

- I am not saving and confirming phone number. 
- Which is the mandatory process to use MFA functionality.
( Multi-Factor Authentication (MFA) increases security for your end users. If you choose 'optional', individual users can have MFA enabled. You can only choose 'required' when initially creating a user pool, and if you do, all users must use MFA. Phone numbers must be verified if MFA is enabled. You can configure adaptive authentication on the Advanced security tab to require MFA based on risk scoring of user sign in attempts.)

Also, I followed docs [here](http://docs.aws.amazon.com/cognito/latest/developerguide/signing-up-users-in-your-app.html) and customised my code to verify both email and phone.

Thanks for the immediate response and your PR is working just fine with the latest version of cognito JS SDK. 

Kudos @hellofornow for the PR

Let me know if you found any solution to add facebook user to user pool :p
Happy weekend!!!

**Comment by joe455** [comment]
@hellofornow  OnSubmit() is not getting called.

Use Submit instead of OnSubmit. 👍

File: src/app/public/auth/mfa/mfa.component.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { Component, Input } from "@angular/core";
+
+@Component({
+    selector: 'awscognito-mfa',
+    templateUrl: './mfa.html'
+})
+export class MFAComponent {
+    @Input() destination: string;
+    @Input() onSubmit: (code: string) => void;
```
Comment (amirtoole) [inline src/app/public/auth/mfa/mfa.component.ts:9]: @joe455 are you referring to this? Not sure why you're referencing `submit`; it should be `onSubmit`. This is passed in from login.html #122

Comment (joe455) [reply src/app/public/auth/mfa/mfa.component.ts:9]: Yes..When I am using OnSubmit with angular4 I am getting error like "OnSubmit is not allowed to bind for security reasons use Submit instead"
So I changed in all three places then it's working perfect!!

I have reproduced error for you..This looks like `below.`

```
Uncaught Error: Template parse errors:
**Binding to event property 'OnSubmit' is disallowed for security reasons, please use (Submit)=**...
If 'OnSubmit' is a directive input, make sure the directive is imported by the current module. ("
        </div>
        <app-mfa [destination]="mfaData.destination"
                 [ERROR ->][OnSubmit]="mfaData.callback">
        </app-mfa>
    
"): ng:///AppModule/LoginComponent.html@50:17
```

**Comment by joe455** [comment]
@hellofornow Need a suggestion from you. I am developing an app with following requirements.

1. My App needs to verify phone and email to activate his account
2. But MFA is required for every user.

Now the problem is, If either email or phone confirmed means, the user state will be CONFIRMED. So,

- I am verifying phone number because otherwise MFA wont work.
- **cognitouser.confirmRegistration** method cant be called for the email verification because user is already CONFIRMED.
- Now if you want to confirm the email the user should be having an active session(I mean logged in) -- So the (1) is not achievable in my point of understanding from [here](http://docs.aws.amazon.com/cognito/latest/developerguide/signing-up-users-in-your-app.html)

Clarify one thing for me...Am I understood it in correct way? @hellofornow

**Comment by joe455** [comment]
Do you have any idea on adding a Facebook user to userpool using Custom UI?
Stuck with it from last 2 weeks (Working on Login with Facebook functionality).

File: src/app/public/auth/mfa/mfa.component.ts (hunk 1 above)
Comment (joe455) [reply src/app/public/auth/mfa/mfa.component.ts]: @hellofornow Got another Issue now.
Once user got registered and not confirmed his phone number. Then, again if he is trying to login I am navigating them to confirmation page(where phone number is not present  so it is undefined). Now in confirmation page the user must enter his phone number but on entering one character the phone number field is disappearing. How can I create work around for this.?

Comment (joe455) [reply src/app/public/auth/mfa/mfa.component.ts]: Hi @hellofornow ,
 Is MFA will work for only phone_numbers. I want users to choose users where they can receive code(email / phone). Can you shed some lite on this?

File: src/app/service/user-login.service.ts [hunk 2]
```diff
@@ -29,40 +57,18 @@ export class UserLoginService {
...
-                let clientParams:any = {};
-                if (environment.sts_endpoint) {
-                    clientParams.endpoint = environment.sts_endpoint;
-                }
-                let sts = new STS(clientParams);
-                sts.getCallerIdentity(function (err, data) {
-                    console.log("UserLoginService: Successfully set the AWS credentials");
-                    callback.cognitoCallback(null, result);
+            newPasswordRequired: (userAttributes, requiredAttributes) => callback.cognitoCallback(`User needs to set password.`, null),
+            onSuccess: result => this.onLoginSuccess(callback, result),
```
Comment (joe455) [inline src/app/service/user-login.service.ts]: onSuccess function is not called

File: src/app/public/auth/mfa/mfa.component.ts (hunk 1 above)
Comment (amirtoole) [reply src/app/public/auth/mfa/mfa.component.ts:9]: I just tested w/ latest version of ng (5.1.0).
I don't think you've copied my code properly.
My version has `[onSubmit]` -- lowercase `o`

Comment (amirtoole) [reply src/app/public/auth/mfa/mfa.component.ts:9]: BTW -- I just pushed my changes which includes updating to latest ng, etc.

Comment (joe455) [reply src/app/public/auth/mfa/mfa.component.ts:9]: ohh..Sorry.
Usually I wont copy code rather will see the flow and will write my own.Might be missed in this way.
Thanks though!!

Comment (joe455) [reply src/app/public/auth/mfa/mfa.component.ts]: @hellofornow Can you please help me to extend my app to use custom authentication flow?
Never mind!!! I am quite new to this.

**Pull request merged**